File: solutions/typescript/square-root/1/square-root.ts
export function squareRoot(radicand: number): number {
  if (radicand < 0) throw new Error("Radicand must be a positive number");
  if (radicand === 0 || radicand === 1) return radicand;

  let left = 1;
  let right = radicand;
  let result = 0;

  while (left <= right) {
    let mid = Math.floor((left + right) / 2);
    let square = mid * mid;

    if (square === radicand) {
      return mid;
    } else if (square < radicand) {
      left = mid + 1;
      result = mid;
    } else {
      right = mid - 1;
    }
  }

  return result;
}

File: solutions/swift/square-root/1/Sources/SquareRoot/SquareRoot.swift
struct SquareRoot {
    static func squareRoot(_ number: Int) -> Int {
        var low = 1
        var high = number
        
        while low <= high {
            let mid = low + (high - low) / 2
            let squared = mid * mid
            
            if squared == number {
                return mid
            } else if squared < number {
                low = mid + 1
            } else {
                high = mid - 1
            }
        }
        
        return high
    }
}
